### PR TITLE
clean up flags in Cache at all costs

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

there were some code paths which could leave the isMigrating() or isResizing() flags set for a Cache object, even when the corresponding MigrateTask or FreeMemoryTask was already finished.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by CryptoNinjaGeek — APPROVED** [review]

File: arangod/Cache/Cache.cpp [hunk 1]
```diff
@@ -431,46 +435,62 @@ bool Cache::migrate(std::shared_ptr<Table> newTable) {
...
+    return true;
+  } catch (...) {
+    if (clearMigrateFlag) {
+      // we must be sure to have cleared the isMigrating() flag here
+      SpinLocker metaGuard(SpinLocker::Mode::Write, _metadata.lock());
+      TRI_ASSERT(_metadata.isMigrating());
+      _metadata.toggleMigrating();
+      TRI_ASSERT(!_metadata.isMigrating());
+    }
+    throw;
```
Comment (MBkkt) [inline arangod/Cache/Cache.cpp:492]: absl::Cleanup (has cancel) or basics::ScopeGuard (not sure but probably also should has cal)

File: arangod/Cache/ManagerTasks.cpp [hunk 2]
```diff
@@ -53,30 +53,56 @@ bool FreeMemoryTask::dispatch() {
...
     _manager.unprepareTask(_environment);
   } catch (...) {
+    if (clearResizingFlag) {
+      // make sure we are always clearing the resizing flag
+      Metadata& metadata = _cache->metadata();
+      SpinLocker metaGuard(SpinLocker::Mode::Write, metadata.lock());
+      TRI_ASSERT(metadata.isResizing());
+      metadata.toggleResizing();
+      TRI_ASSERT(!metadata.isResizing());
+    }
```
Comment (MBkkt) [inline arangod/Cache/ManagerTasks.cpp:104]: NIT
cleanup

File: arangod/Cache/Cache.cpp [hunk 3]
```diff
@@ -183,17 +187,17 @@ bool Cache::isMigrating() const noexcept {
...
 
-bool Cache::isBusy() const noexcept {
-  if (ADB_UNLIKELY(isShutdown())) {
-    return false;
-  }
-
+bool Cache::isMigratingOrResizingFlagSet() const noexcept {
   SpinLocker metaGuard(SpinLocker::Mode::Read, _metadata.lock());
-  return _metadata.isResizing() || _metadata.isMigrating();
+  return _metadata.isMigrating() || _metadata.isResizing();
```
Comment (MBkkt) [inline arangod/Cache/Cache.cpp:200]: Why reorder?

**Review by MBkkt — CHANGES_REQUESTED** [review]
Avoiding catch exception is good practice for performance critical code.
In this code it can be avoided with raii
https://godbolt.org/z/W4bMEfser

**Review by MBkkt — APPROVED** [review]

**Pull request merged**